src/openrct2/management/Research.cpp

```diff
@@ -151,33 +151,16 @@ static void ResearchNextDesign()
         return;
     }
 
-    ResearchItem researchItem;
-
-    bool ignoreActiveResearchTypes = false;
-    auto it = gResearchItemsUninvented.begin();
-    for (;;)
+    // Try to find a research item of a matching type, if none found, use any first item
+    auto it = std::find_if(gResearchItemsUninvented.begin(), gResearchItemsUninvented.end(), [](const auto& e) {
+        return (gResearchPriorities & EnumToFlag(e.category)) != 0;
+    });
+    if (it == gResearchItemsUninvented.end())
     {
-        researchItem = *it;
-        if (it == gResearchItemsUninvented.end())
-        {
-            if (!ignoreActiveResearchTypes)
-            {
-                ignoreActiveResearchTypes = true;
-                it = gResearchItemsUninvented.begin();
-                continue;
-            }
-
-            ResearchMarkAsFullyCompleted();
-            return;
-        }
-
-        if (ignoreActiveResearchTypes || (gResearchPriorities & EnumToFlag(researchItem.category)))
-        {
-            break;
-        }
-        it++;
+        it = gResearchItemsUninvented.begin();
     }
 
+    ResearchItem researchItem = std::move(*it);
```

Comment:
ResearchItem's are tiny objects just copy them. Also not sure it should be a move anyway.

Reply:
Before my PR they were being moved already, and there's no harm in using `std::move` if it's a no-op.

Reply:
The previous code moved the `reasearchItem` but it didn't move from the iterator it copied. Does it make a difference, no probably not. But it could cause issues if we weren't about to delete the iterators value. (also in MSVC std::move isn't a no-op in debug builds {but it is now in the very latest MSVC})

Reply:
Good point about MSVC, changed.